department-of-veterans-affairs/va.gov-cms
# VACMS-3378: Selecting and unselecting "Reusable" shows and hides the "Save and select" button...

…in the media library dialog.

## Description

See #3378.

This is ugly.

As documented in #3378, I attempted to solve this using the official solution of the Forms API `#states` system, but that clashes horribly with the Core `dialog` system and both seem at least a little broken individually.  I attempted yesterday to fix the intersection between the two, and realized after some time that I'm not the man to do it.  At least, not yet.

The solution I ended up going with took two parts:
- added a custom class to the `field_media_in_library` checkbox widget to make it easier for jQuery to identify cleanly.  I could've used regex tests but I felt this might be cleaner and more transparent.
- attached a custom library to the media upload form which scans for appearances of that specific checkbox widget and listens to their events; when the checkbox is unchecked, it hides the dialog button, and when the checkbox is checked, it shows the dialog button.  Additionally, it will check the existing state of the checkbox and set the visibility of the button accordingly, which is significant because reloads of the form (for instance, in response to errors) can present the checkbox and button in mismatched states.

I bet there's a better way to do this, and I'm completely open to suggestions/corrections/pitchfork-shaking demands.  If there is, I couldn't figure it out 😩 

## Testing done
Repeatedly tried to add media, clicked "Reusable" a few times, added media with existing media, etc.  No automated tests yet.  I think that would be a good idea, so... forthcoming?

## Screenshots
<img width="972" alt="Screen Shot 2020-12-02 at 12 38 13 PM" src="https://user-images.githubusercontent.com/1318579/100909963-57a5f000-349b-11eb-826c-8086bd52f1d6.png">
<img width="949" alt="Screen Shot 2020-12-02 at 12 38 20 PM" src="https://user-images.githubusercontent.com/1318579/100909968-5a084a00-349b-11eb-9931-ef056d623670.png">

## QA steps
As a user with the ability to edit nodes containing media content blocks:
1. Edit [a node that can have attached media](http://va-gov-cms.lndo.site/node/7016/edit)
1. Click "Add Content block".
1. Select either "Embedded Image" or "Link to File or Video".
1. Click "Add media".
1. Click "Choose file" and upload a file from your machine.
1. Then validate Acceptance Criteria from #3378:
- [ ] the "Save and Select" button should be visible when "Reusable" is checked
- [ ] the "Save and Select" button should be invisible when "Reusable" is unchecked

### Definition of Done


### Select Team for PR review


### Does this PR need review from a Product Owner


### CMS user-facing announcement
This is going to simplify the UX, so I assume this will need an announcement.

Is an announcement needed to let editors know of this change?
